### ISSUE TITLE
Cannot reproduce ADE20k

### ISSUE THREAD

**Issue by Ze-Yang** [HieuPhan33/REMINDER#2]
Hi, I tried to reproduce the ADE20k results as below. I reuse the step 0 checkpoint I train with PLOP (mIoU 41.98), as I think REMINDER shares the same training setup with PLOP for step 0. I attach the script [reminder_100-50.sh](https://github.com/HieuPhan33/REMINDER/files/9332629/reminder_100-50.sh.txt) and [reminder_100-10.sh](https://github.com/HieuPhan33/REMINDER/files/9332626/reminder_100-10.sh.txt) I use for your reference. I notice that you use the linear lr scheduling strategy when you scale down the batch size to 10 per GPU. For consistency, I adopt this strategy across all settings with the lr 0.0008. Additionally, I train the script with 2 GPUs as PLOP does. Hope for your advice.

|            | 100-50 |       |      | 100-10 |       |      | 50-50 |       |      |
|------------|:------:|:-------:|:------:|:------:|:-------:|:------:|:------:|:-------:|:------:|
|            | 0-100 | 101-150 |  all | 0-100 | 101-150 |  all | 0-50 | 51-150 |  all |
| Reported   | 41.55 |  **19.16** | **34.14** | **38.96** | **21.28** | **33.11** | **47.11** |  **20.35** | **29.39** |
| Reproduced | **41.91** | 16.10 | 33.36 | 38.43 | 15.56 | 30.86 | 45.56 | 18.49 | 27.63 |

**Comment by HieuPhan33** [comment]
I find that training on a longer number of epochs could help. You can try setting the #epochs to 80.

FYI, I share with the my results:
[2021-09-13_ade_100-5_REMINDER.csv](https://github.com/HieuPhan33/REMINDER/files/9334485/2021-09-13_ade_100-5_REMINDER.csv)
[2022-03-26_ade_100-10_REMINDER.csv](https://github.com/HieuPhan33/REMINDER/files/9334486/2022-03-26_ade_100-10_REMINDER.csv)
[2022-05-23_ade_100-50_REMINDER.csv](https://github.com/HieuPhan33/REMINDER/files/9334487/2022-05-23_ade_100-50_REMINDER.csv)

For 50-50, our performance is not consistent. I'm trying to reproduce the results, and will send you the log result later.

**Comment by Ze-Yang** [comment]
I notice that the prototype loss and update after each step do not support multiple GPUs, as the prototypes are not `reduce` or `gather` (see DDP operations) to a single GPU.

**Comment by Ze-Yang** [comment]
Hi @HieuPhan33, I can reproduce the results now. The only change is that I run with a single GPU this time. It validates my thought that the prototype computation does not support multiple GPUs since not `reduce` or `gather`. I suggest that if this issue is fixed, the performance on VOC will be higher since it currently runs with the sub-batch prototypes (more specifically half batch with 2 GPUs).

|            | 100-50 |       |      | 100-10 |       |      | 50-50 |       |      |
|------------|:------:|:-------:|:------:|:------:|:-------:|:------:|:------:|:-------:|:------:|
|            | 0-100 | 101-150 |  all | 0-100 | 101-150 |  all | 0-50 | 51-150 |  all |
| Reported   | 41.55 | 19.16 | 34.14 | 38.96 | 21.28 | 33.11 | 47.11 |  20.35 | 29.39 |
| Reproduced | 41.85 | 22.29 | 35.37 | 38.71 | 20.33 | 32.63 | 47.56 | 20.12 | 29.39 |

**Comment by HieuPhan33** [comment]
Oh, thanks @Ze-Yang, that's a great contribution.

I will try training Pascal-VOC using a single GPU and update the latest result on the Github repo.
